src/t_hash.c

```diff
@@ -759,22 +785,69 @@ sds hashTypeGetFromHashTable(robj *o, sds field) {
  * can always check the function return by checking the return value
  * for C_OK and checking if vll (or vstr) is NULL. */
 int hashTypeGetValue(robj *o, sds field, unsigned char **vstr, unsigned int *vlen, long long *vll) {
+    sds key;
+    GetFieldRes res;
     if (o->encoding == OBJ_ENCODING_LISTPACK ||
-        o->encoding == OBJ_ENCODING_LISTPACK_EX)
-    {
+        o->encoding == OBJ_ENCODING_LISTPACK_EX) {
         *vstr = NULL;
-        if (hashTypeGetFromListpack(o, field, vstr, vlen, vll) == 0)
+        res = hashTypeGetFromListpack(o, field, vstr, vlen, vll);
+
+        if (res == GET_FIELD_OK)
             return C_OK;
+
+        if (res == GET_FIELD_NOT_FOUND)
+            return C_ERR;
+
+        /* If reached here then res == GET_FIELD_EXPIRED. Extract key. */
+        serverAssert(o->encoding == OBJ_ENCODING_LISTPACK_EX);
+        listpackEx *lpt = o->ptr;
+        key = lpt->key;
+
     } else if (o->encoding == OBJ_ENCODING_HT) {
-        sds value;
-        if ((value = hashTypeGetFromHashTable(o, field)) != NULL) {
+        sds value = NULL;
+        res = hashTypeGetFromHashTable(o, field, &value);
+        if (res == GET_FIELD_OK) {
             *vstr = (unsigned char*) value;
             *vlen = sdslen(value);
             return C_OK;
         }
+
+        if (res == GET_FIELD_NOT_FOUND)
+            return C_ERR;
+
+        /* If reached here then res == GET_FIELD_EXPIRED. Extract key. */
+        serverAssert(isDictWithMetaHFE((dict*)o->ptr));
+        key = ((dictExpireMetadata *) dictMetadata((dict*)o->ptr))->key;
+
     } else {
         serverPanic("Unknown hash encoding");
     }
+
+    /* If in replica then return C_OK (discard hash-field expiry logic) */
+    if ((server.masterhost) && (server.current_client && (server.current_client->flags & CLIENT_MASTER)))
+        return C_OK;
+
+    /*  TODO TODO TODO TODO TODO TODO TODO TODO TODO TODO TODO TODO TODO
+     * Once dbid will be embedded inside HASH, extract it from there and remove
+     * this hack that is aligned with testing.
+     *
+     * Going to be my next commit.
+     */
+    redisDb *db = server.db+0;
+
+    /* delete the field and propagate the deletion */
+    serverAssert(hashTypeDelete(o, field, 1) == 1);
+    propagateHashFieldDeletion(db, key, field, sdslen(field));
+
+    if (hashTypeLength(o, 0) == 0) {
+
+        /* create temporary robj with the key */
+        robj *keyObj = createStringObject(key, sdslen(key));
+
+        notifyKeyspaceEvent(NOTIFY_GENERIC, "del", keyObj, db->id);
+        dbDelete(db,keyObj);
+        decrRefCount(keyObj);
+    }
```

Comment:
is it more readable to move these code that is executed for replication info if?
like:
```c
    /* If in replica then return C_OK (discard hash-field expiry logic) */
    if (!(server.masterhost) || !(server.current_client && (server.current_client->flags & CLIENT_MASTER))) {
        ... do things
        return C_ERR;
    }
```

Reply:
Please consider my last commit